contracts/zero-ex/contracts/src/features/TransformERC20.sol

```diff
@@ -0,0 +1,365 @@
+        Transformation[] memory transformations
+    )
+        public
+        override
+        payable
+        onlySelf
+        returns (uint256 outputTokenAmount)
+    {
+        return _transformERC20Private(
+            callDataHash,
+            taker,
+            inputToken,
+            outputToken,
+            inputTokenAmount,
+            minOutputTokenAmount,
+            transformations
+        );
+    }
+
+    /// @dev Private version of `transformERC20()`.
+    /// @param callDataHash Hash of the ingress calldata.
+    /// @param taker The taker address.
+    /// @param inputToken The token being provided by the taker.
+    ///        If `0xeee...`, ETH is implied and should be provided with the call.`
+    /// @param outputToken The token to be acquired by the taker.
+    ///        `0xeee...` implies ETH.
+    /// @param inputTokenAmount The amount of `inputToken` to take from the taker.
+    /// @param minOutputTokenAmount The minimum amount of `outputToken` the taker
+    ///        must receive for the entire transformation to succeed.
+    /// @return outputTokenAmount The amount of `outputToken` received by the taker.
+    function _transformERC20Private(
+        bytes32 callDataHash,
+        address payable taker,
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        uint256 minOutputTokenAmount,
+        Transformation[] memory transformations
+    )
+        public
+        payable
+        returns (uint256 outputTokenAmount)
+    {
+        // If the input token amount is -1, transform the taker's entire
+        // spendable balance.
+        if (inputTokenAmount == uint256(-1)) {
+            inputTokenAmount = ITokenSpender(address(this))
+                .getSpendableERC20BalanceOf(inputToken, taker);
+        }
+
+        IPuppet puppet = getTransformPuppet();
+
+        // Remember the initial output token balance of the taker.
+        uint256 takerOutputTokenBalanceBefore = minOutputTokenAmount == 0
+            ? 0 : LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+
+        // Pull input tokens from the taker to the puppet and transfer attached ETH.
+        _transferInputTokensAndAttachedEth(inputToken, taker, address(puppet), inputTokenAmount);
+
+        // Perform transformations.
+        for (uint256 i = 0; i < transformations.length; ++i) {
+            _executeTransformation(puppet, transformations[i], taker, callDataHash);
+        }
+
+        // Ensure enough output token has been sent to the taker.
+        if (minOutputTokenAmount != 0) {
+            // Compute how much output token has been transferred to the taker.
+            uint256 takerOutputTokenBalanceAfter =
+                LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+            if (takerOutputTokenBalanceAfter > takerOutputTokenBalanceBefore) {
+                outputTokenAmount = takerOutputTokenBalanceAfter.safeSub(
```

Comment:
It would seem to me that we would always want to compute `outputTokenAmount`, no?

Reply:
It'll stay at zero if the balance does not increase. We can't represent decreases with a uint256.

Reply:
For some more context, suppose `minOutputTokenAmount == 0`. Then the user doesn't mind how much they receive, but wouldn't we still want to emit and return the value? Additionally, if `takerOutputTokenBalanceAfter < takerOutputTokenBalanceBefore` then the `safeSub` will just revert.

Reply:
Yeah, that makes sense. We probably still want to handle a negative change because SafeMath rich reverts are obnoxiously opaque.

Reply:
I kinda like having the option of not querying balances twice, e.g. for (un)wrapping WETH it'd be a waste of gas to do so

Reply:
I do too, but I also think it's pretty useful for this function to return the amount transferred regardless.